src/main/kotlin/no/nav/klage/oppgave/service/VedtakService.kt

```diff
@@ -158,22 +148,86 @@ class VedtakService(
-        val klage = klagebehandlingRepository.findById(klagebehandlingId).orElseThrow()
-        val vedtak = klage.vedtak.find { it.id == vedtakId }
-        require(vedtak != null) { "Fant ikke vedtak på klage" }
+    fun finalizeVedtak(
+        klagebehandlingId: UUID,
+        vedtakId: UUID,
+        input: VedtakFullfoerInput,
+        innloggetIdent: String
+    ) {
+        val klage = klagebehandlingService.getKlagebehandlingForUpdate(
+            klagebehandlingId,
+            input.klagebehandlingVersjon
+        )
+        val vedtak = getVedtakFromKlagebehandling(klage, vedtakId)
+        if (vedtak.finalized != null) throw VedtakFinalizedException("Vedtak med id $vedtakId er allerede ferdigstilt")
+        if (vedtak.journalpostId == null) throw JournalpostNotFoundException("Vedtak med id $vedtakId er ikke journalført")
         require(vedtak.utfall != null) { "Utfall på vedtak må være satt" }
-        require(vedtak.journalpostId != null) { "Kan ikke fullføre et vedtak uten journalført vedtaksbrev" }
-        val utfall = vedtak.utfall!!
-        val vedtakFattet = KlagevedtakFattet(
-            kildeReferanse = klage.kildeReferanse ?: "UKJENT",
-            kilde = klage.kildesystem.name,
-            utfall = utfall,
-            vedtaksbrevReferanse = vedtak.journalpostId,
-            kabalReferanse = vedtakId.toString()
+
+        finalizeJournalpost(
+            klage,
+            vedtak,
+            innloggetIdent,
+            input.journalfoerendeEnhet
         )
 
-        vedtakKafkaProducer.sendVedtak(vedtakFattet)
+        addVedtakToStore(klage, vedtak)
+    }
+
+    private fun finalizeJournalpost(
+        klagebehandling: Klagebehandling,
+        vedtak: Vedtak,
+        utfoerendeSaksbehandlerIdent: String,
+        journalfoerendeEnhet: String
+    ): Vedtak? {
+        return try {
+            joarkClient.finalizeJournalpost(vedtak.journalpostId!!, journalfoerendeEnhet)
+            setFinalized(klagebehandling, vedtak.id, utfoerendeSaksbehandlerIdent)
+        } catch (e: Exception) {
+            logger.warn("Kunne ikke ferdigstille journalpost ${vedtak.journalpostId}")
+            throw JournalpostFinalizationException("Klarte ikke å journalføre vedtak")
+        }
+    }
+
+    // TODO Legg til brevutsending i eget steg?
+
+    private fun addVedtakToStore(klage: Klagebehandling, vedtak: Vedtak) {
+        kafkaVedtakEventRepository.save(
+            KafkaVedtakEvent(
+                kildeReferanse = klage.kildeReferanse ?: "UKJENT",
+                kilde = klage.kildesystem.name,
+                utfall = vedtak.utfall!!,
+                vedtaksbrevReferanse = vedtak.journalpostId,
+                kabalReferanse = vedtak.id.toString(),
+                status = UtsendingStatus.IKKE_SENDT
+            )
+        )
+    }
+
+    @Scheduled(cron = "0 0 3 * * *", zone = "Europe/Paris")
+    private fun dispatchUnsendtVedtakToKafka() {
```

Comment:
Jeg tror jeg savner en @Transactional-annotasjon her. Lurer på om du må gjøre metoden public også da, for at Spring skal klare å instrumentere dette skikkelig. Og muuuuuuligens flytte @Scheduled ut av denne klassen, men jeg er ikke sikker på det siste der.